pkg/filter/network/tunnel/agent.go

```diff
@@ -0,0 +1,183 @@
+	"net"
+	"sync"
+	"time"
+
+	v2 "mosn.io/mosn/pkg/config/v2"
+	"mosn.io/mosn/pkg/filter/network/tunnel/ext"
+	"mosn.io/mosn/pkg/log"
+	"mosn.io/mosn/pkg/server"
+	"mosn.io/pkg/utils"
+)
+
+type agentBootstrapConfig struct {
+	Enable bool `json:"enable"`
+	// The number of connections established between the agent and each server
+	ConnectionNum int `json:"connection_num"`
+	// The cluster of remote server
+	Cluster string `json:"cluster"`
+	// After the connection is established, the data transmission is processed by this listener
+	HostingListener string `json:"hosting_listener"`
+	// Static remote server list
+	StaticServerList []string `json:"server_list"`
+
+	// DynamicServerListConfig is used to specify dynamic server configuration
+	DynamicServerListConfig struct {
+		DynamicServerLister string `json:"dynamic_server_lister"`
+	}
+
+	// ConnectRetryTimes
+	ConnectRetryTimes int `json:"connect_retry_times"`
+	// ReconnectBaseDuration
+	ReconnectBaseDurationMs int `json:"reconnect_base_duration_ms"`
+
+	// ConnectTimeoutDurationMs specifies the timeout for establishing a connection and initializing the agent
+	ConnectTimeoutDurationMs int    `json:"connect_timeout_duration_ms"`
+	CredentialPolicy         string `json:"credential_policy"`
+}
+
+func init() {
+	v2.RegisterParseExtendConfig("tunnel_agent", func(config json.RawMessage) error {
+		var conf agentBootstrapConfig
+		err := json.Unmarshal(config, &conf)
+		if err != nil {
+			log.DefaultLogger.Errorf("[tunnel agent] failed to parse agent bootstrap config: %v", err.Error())
+			return err
+		}
+		if conf.Enable {
+			utils.GoWithRecover(func() {
+				bootstrap(&conf)
+			}, nil)
+
+		}
+		return nil
+	})
+}
+
+func bootstrap(conf *agentBootstrapConfig) {
+	if conf.DynamicServerListConfig.DynamicServerLister != "" {
+		utils.GoWithRecover(func() {
+			lister := ext.GetServerLister(conf.DynamicServerListConfig.DynamicServerLister)
+			ch := lister.List(conf.Cluster)
+			for {
+				select {
+				case servers := <-ch:
+					// Compute the diff between new and old server list
+					intersection := make(map[string]bool)
+					for i := range servers {
+						if _, ok := connectionMap.Load(servers[i]); ok {
+							intersection[servers[i]] = true
+						}
+					}
+					increased := make([]string, 0)
+					for _, addr := range servers {
+						if _, ok := intersection[addr]; !ok {
+							increased = append(increased, addr)
+							utils.GoWithRecover(func() {
+								connectServer(conf, addr)
+							}, nil)
+						}
+					}
+					decreased := make([]string, 0)
+					connectionMap.Range(func(key, value interface{}) bool {
+						addr := key.(string)
+						_, ok := intersection[addr]
+						if !ok {
+							decreased = append(decreased, addr)
+						}
+						return true
+					})
+					for _, addr := range decreased {
+						val, ok := connectionMap.LoadAndDelete(addr)
+						if !ok {
+							continue
+						}
+						for _, conn := range val.([]*AgentRawConnection) {
+							err := conn.Stop()
```

Comment:
support graceful stop

Reply:
TODO？

Reply:
it's easy to support it， and you could call RemoveClusterHosts and ShutdownConnectionPool when graceful
stop.